amazon_transcribe/serialize.py

```diff
@@ -12,75 +12,119 @@
 
 
-from io import BufferedIOBase
-from typing import Dict, Tuple, Union
+from typing import Dict, Tuple, Optional
 
-from amazon_transcribe.request import PreparedRequest, Request
+from amazon_transcribe.request import Request
 from amazon_transcribe.structures import BufferableByteStream
 from amazon_transcribe.utils import _add_required_headers
-from amazon_transcribe.exceptions import SerializerException
+from amazon_transcribe.exceptions import SerializerException, ValidationException
 from amazon_transcribe.model import (
     AudioEvent,
     StartStreamTranscriptionRequest,
     BaseEvent,
 )
 
-HEADER_VALUE = Union[int, None, str]
+HEADER_PREFIX = "x-amzn-transcribe-"
 
 
-class Serializer:
-    def __init__(self):
-        raise NotImplementedError("Serializer")
-
-    def serialize(self) -> Tuple[Dict[str, HEADER_VALUE], BufferedIOBase]:
-        """Serialize out to payload and headers."""
-        raise NotImplementedError("serialize")
-
-    def serialize_to_request(self) -> PreparedRequest:
-        """Serialize parameters into an HTTP request."""
-        raise NotImplementedError("serialize_to_request")
-
-
-class TranscribeStreamingRequestSerializer(Serializer):
+class TranscribeStreamingSerializer:
     """Convert StartStreamTranscriptionRequest into a
     Request object for streaming to the Transcribe service.
     """
 
-    def __init__(self, endpoint, transcribe_request):
-        self.endpoint: str = endpoint
-        self.method: str = "POST"
-        self.request_uri: str = "/stream-transcription"
-        self.request_shape: StartStreamTranscriptionRequest = transcribe_request
+    def _raise_param_validation_exception(self, header, expected):
+        raise ValidationException(
+            f"Unexpected param type encountered for {header}, expected type: {expected}"
+        )
 
-    def serialize(self) -> Tuple[Dict[str, HEADER_VALUE], BufferedIOBase]:
-        headers = {
-            "x-amzn-transcribe-language-code": self.request_shape.language_code,
-            "x-amzn-transcribe-sample-rate": self.request_shape.media_sample_rate_hz,
-            "x-amzn-transcribe-media-encoding": self.request_shape.media_encoding,
-            "x-amzn-transcribe-vocabulary-name": self.request_shape.vocabulary_name,
-            "x-amzn-transcribe-session-id": self.request_shape.session_id,
-            "x-amzn-transcribe-vocabulary-filter-method": self.request_shape.vocab_filter_method,
-            "x-amzn-transcribe-vocabulary-filter-name": self.request_shape.vocab_filter_name,
-            "x-amzn-transcribe-show-speaker-label": self.request_shape.show_speaker_label,
-            "x-amzn-transcribe-enable-channel-identification": self.request_shape.enable_channel_identification,
-            "x-amzn-transcribe-number-of-channels": self.request_shape.number_of_channels,
-        }
+    def _serialize_str_header(self, header: str, value: Optional[str]):
+        if value is None:
```

Comment:
Given these are all almost identical, would it be better to have a generic `_serialize_header` that takes a type? We can call that from here and then expand these later if we have custom logic required for a given type.

Reply:
I think that makes sense so long as it doesn't remove information from the typing system and the thing being serialized is handled correctly. For example, I'm not sure if a parameter that's required should be dropped if it's `None`. For now I suppose it's fine as that was the existing behavior and you'll probably just end up getting an exception from the service but I think this might be something we want to take a deeper dive on in the future.